### PR TITLE
Cleanup from lexical MIR borrowck removal

### PR DESCRIPTION
**Pull request by matthewjasper**
Lexical MIR borrowck was removed months ago now, and `EndRegion`s are no longer used for MIRI verification.

* Remove `rustc::mir::StatementKind::EndRegion` and the `-Zemit_end_regions` flag
* Use `RegionVid` instead of `Region` in BorrowSet
* Rewrite drop generation to create fewer goto terminators.

r? @nikomatsakis

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
The last commits sound potentially perf-sensitive, so @bors try

**Comment by kennytm** [comment]
@rust-timer build 5247df0b727520b9128fbf5b55797a25859a3dd8

**Comment by kennytm** [comment]
Looks pretty green to me except serde-check which is +1.4%

**Comment by matthewjasper** [comment]
Added a somewhat speculative performance fix, but it looks OK locally.
@bors try

**Comment by matthewjasper** [comment]
@rust-timer build 6f6b42de19da32198b926c7eefd817e4ada47ca2

**Comment by kennytm** [comment]
@rust-timer build 6f6b42de19da32198b926c7eefd817e4ada47ca2

File: src/librustc_mir/dataflow/impls/borrows.rs [hunk 1]
```diff
@@ -349,52 +321,7 @@ impl<'a, 'gcx, 'tcx> BitDenotation for Borrows<'a, 'gcx, 'tcx> {
...
-            panic!("could not find block at location {:?}", location);
-        });
-
-        let term = block.terminator();
-        match term.kind {
-            mir::TerminatorKind::Resume |
-            mir::TerminatorKind::Return |
-            mir::TerminatorKind::GeneratorDrop => {
-                // When we return from the function, then all `ReScope`-style regions
-                // are guaranteed to have ended.
```
Comment (nikomatsakis) [inline src/librustc_mir/dataflow/impls/borrows.rs]: nice, lots of dead code

**Review by nikomatsakis — APPROVED** [review]
r=me but tests are failing for some reason?

**Comment by matthewjasper** [comment]
Looks like the performance issue was elsewhere as well.

**Comment by matthewjasper** [comment]
OK, so it seems that generating drops in reverse order results in various MIR passes taking longer. Drops are now back to being generated in execution order (except for unwind cleanup).

@bors try

**Pull request merged**